Allow train/eval, and non-Tensor arguments to python functions

zdevito opened the pull request:
This whitelists train/eval functions in script modules, and tests that nested nn.Modules still work.

This also changes the code for calling python functions from script to allow non-tensor inputs/outputs.

zdevito committed:
Whitelist train/eval

zdevito committed:
Support non-tensor arguments/returns to python functions.

apaszke commented on test/test_jit.py:
```diff
@@ -457,6 +457,52 @@ def method(self, x):
+                # TODO: it is illegal to try to call
+                # eval/train because training has already
+                # been set. Consider allowing
+                # constants to be mutable until the end of __init__
```
Well if that's the case then having it be a constant is a bit useless. Pretty much every `not self.training` branch is guaranteed to always be dead code.

zdevito replied:
Yes, it is useless, but I wanted to make sure it was consistent with how constants behaved at least.  In the future, we should support non-tensor module parameters so that training can be an actual parameter.

apaszke commented on test/test_jit.py:
```diff
@@ -4400,6 +4446,32 @@ def forward(self, x, seq_lens):
+    def test_python_call_non_tensor(self):
+        def foo(a, b, c):
+            d, e = c
+            # type: (Tensor, int, Tuple[Tensor, int]) -> Tuple[int, Tensor]
```
Doesn't this need to be on the first line?

zdevito replied:
I checked and it doesn't actually matter. I think the regex doesn't actually check that it has to be the first line. I changed the test regardless.

apaszke commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -167,6 +167,27 @@ inline IValue argumentToIValue(
+    const TypePtr& type,
+    py::handle object) {
+  try {
+    return toIValue(object, type);
```
Couldn't we replace all uses of this function with this/a similar helper? Our error messages are pretty bad at the moment.

apaszke commented on torch/csrc/jit/python_interpreter.cpp:
```diff
@@ -45,9 +46,7 @@ Operation createPythonOperation(Node* op_) {
         py_inputs[i] = py::reinterpret_borrow<py::object>(
             op->scalar_args[next_scalar++].get());
```
Do we still use those static arguments for anything? It would be better to embed them as constants in the graph

zdevito replied:
I think they are still used for calls to python Function objects in the tracer.

apaszke commented on torch/csrc/jit/python_interpreter.cpp:
```diff
@@ -45,9 +46,7 @@ Operation createPythonOperation(Node* op_) {
         py_inputs[i] = py::reinterpret_borrow<py::object>(
             op->scalar_args[next_scalar++].get());
       } else if (arg_type == 't') {
```
It would be nice to rename `'t'` to `'d'` as in dynamic instead of tensor.

apaszke commented on torch/csrc/jit/python_interpreter.cpp:
```diff
@@ -56,30 +55,25 @@ Operation createPythonOperation(Node* op_) {
+      stack.push_back(returnToIValue(pos, op->outputs().at(pos)->type(), entry));
     };
 
     if (!PyTuple_Check(py_outputs.ptr())) {
```
Do we treat the `-> Tuple[int, Tensor]` schema differently from `-> (int, Tensor)`? If yes, I think that we should branch on `num_outputs` instead of checking if the return value is a tuple. This would be a good test too.

zdevito replied:
yeah, that is a better test. Ill make that change.

apaszke reviewed (commented):
Script module LGTM.
PythonOp code looks fishy.

zdevito committed:
Change PythonOp so it always returns a single value (that may be a tuple)

zdevito committed:
improve cquery results

zdevito committed:
simplify and fix minor bugs in error reporting

zdevito commented:
updated so that Python calls always return a single (possibly tuple) value, which clears up the tuple confusion. I also simplified the error message reporting and made sure it doesn't return stacktraces of C++ code for user errors.

apaszke commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -26,22 +26,22 @@ inline void findErrorInKwargs(
-        schema.name, "'. Schema: ", schema);
+            [&key](const Argument& argument) { return argument.name == key; })) {
+        throw std::runtime_error(at::str("Unknown keyword argument '", key, "' for operator '",
+        schema.name, "'. Schema: ", schema));
```
If the problem here was with `AT_CHECK`, then we should probably add some comments. Otherwise, sooner or later, someone will codemod those to become `AT_*` macros again. 😕 Ideally we'd modify pybind to know that it shouldn't display those messages. `HANDLE_TH_ERRORS` macros disable them.

apaszke commented on torch/csrc/jit/python_ir.cpp:
```diff
@@ -180,29 +180,6 @@ void initPythonIRBindings(PyObject * module_) {
        py::arg("defer_weight_export")=false,
        py::arg("operator_export_type")=::torch::onnx::OperatorExportTypes::ONNX,
        py::arg("google_printer")=false)
-    .def("wrapPyFuncWithSymbolic", [](Graph &g, py::function func, std::vector<Value*> inputs, size_t n_outputs, py::function symbolic) {
```
😍

apaszke commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -69,18 +69,15 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
-      std::vector<TypePtr> arg_types, ret_types;
-      std::tie(arg_types, ret_types) = py::cast<std::pair<std::vector<TypePtr>, std::vector<TypePtr>>>(signature);
+      std::vector<TypePtr> arg_types;
+      TypePtr ret_type;
```
😍

apaszke commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -69,18 +69,15 @@ struct VISIBILITY_HIDDEN PythonValue : public SugaredValue {
-      for (auto &ret_type : ret_types) {
-        rets.push_back(Argument(std::to_string(idx++), std::move(ret_type), {}, {}, false));
-      }
+      rets.push_back(Argument(std::to_string(0), std::move(ret_type), {}, {}, false));
```
nit: `std::to_string(0)` is `"0"` 😉

apaszke reviewed (approved):
Nice cleanups! Just to clarify: this means that we treat `-> (...)` and `-> Tuple[...]` as the same return type, right?

zdevito commented:
yes, `(a, b)` in a return gets parsed as a tuple, just like `Tuple[a, b]`. Eventually we should also change user-defined script functions to also return single arguments, but I held off because this would interact poorly with onnx.

zdevito committed:
PR comments
